BUG: random: Don't return negative values from Generator.geometric.

WarrenWeckesser opened the pull request:
In C, when the integer part of a double exceeds the maximum int64, casting the double to int64_t is undefined behavior.  On some platforms, the result is `2**64-1` and on others it is `-2**64`.  That results in rng.geometric() returning `-2**64` when p was sufficiently small on some platforms.

The fix is to never invoke undefined behavior by ensuring we don't attempt to cast very large double values to int64_t.

See gh-17007.

WarrenWeckesser committed:
BUG: random: Don't return negative values from Generator.geometric.

In C, when the integer part of a double exceeds the maximum int64, casting
the double to int64_t is undefined behavior.  On some platforms, the result
is 2**64-1 and on others it is -2**64.  That results in rng.geometric()
returning -2**64 when p was sufficiently small on some platforms.

The fix is to never invoke undefined behavior by ensuring we don't attempt
to cast very large double values to int64_t.

rkern commented on numpy/random/src/distributions/distributions.c:
```diff
@@ -960,7 +960,15 @@ RAND_INT_TYPE random_geometric_search(bitgen_t *bitgen_state, double p) {
+   * The constant 9.223372036854776e+18 is the smallest double that is
+   * larger than INT64_MAX.
+   */
+  if (z >= 9.223372036854776e+18) {
```
Why not `z > (double)INT64_MAX`?

WarrenWeckesser replied:
`(double)INT64_MAX` and `9.223372036854776e+18` are the same value, so yeah, that spelling also works.  The comparison must be `>=`, though, because the value of `(double)INT64_MAX` is actually `INT64_MAX + 1` (i.e. `2**63`) and we don't want to allow that value to get through and invoke undefined behavior.

rkern replied:
Ah, right, because of the spacing. The literal is probably clearer than the comment required for `z >= (double)INT64_MAX` would be.

WarrenWeckesser replied:
By "the literal", do you mean `9.223372036854776e+18`?

rkern replied:
Yes.

rkern reviewed (approved):
LGTM. The failures seem unrelated.

I'm a little surprised that that regressions test suite is using a non-deterministically seeded shared `Generator` instead of one created deterministically in a `setUp`. If you wanted to drive-by fix that, I think it would be welcome, but not necessary.

WarrenWeckesser commented:
I'll add the setup code to the test class.

WarrenWeckesser committed:
MAINT: random: In a test module, don't use a global non-deterministically Generator.

WarrenWeckesser commented:
I added `setup_method()` to the `TestRegression` class and removed the global `mt19937`.  Some of the methods in the class create seeded instances of `Generator` locally so they can check for reproducible results.  I didn't touch those.

seberg commented:
Thanks Warren and Robert.  I suspect coverage is missing things because the compiler inlined the function.
